doc/event-alarm-framework/events-producer.md

```diff
@@ -0,0 +1,212 @@
+```
+module events-bgp {
+    . . .
+    container bgp_status {
+        list event_list {
+            key "type ip"
+            
+            leaf type {
+                enum "admin_up";
+                enum "admin_down";
+                enum "idle";
+                enum "active";
+                enum "open";
+                enum "established";
+            }
+            
+            leaf ip {
+                type inet:ip-address;
+                description "IP of neighbor";
+            }
+            
+            leaf timestamp {
+                type inet::date-and-time;
+                description "time of the event"
+            }
+        }
+    }
+``` 
+
+### Event detection
+The event detection could happen in many ways
+- Update the code to directly invoke Event reporter, which will stream it out.
+- Run a periodic check (via monit/cron/daemon/...) to identify an event and raise it via event reporter
+- Watch syslog messages for specific pattern; Upon match parse the required data out and call event reporter.
+- Any other means of getting alerted on an event
+
+#### Log message based detection
+At high level:
+1. This is a two step process.
+2. The process raising the event sends a sylog message out.
+3. A watcher scans all the syslog messages emitted and parse/check for events of interest.
+4. When matching message arrives, raise the event
+
+Here you have code that sends the log and a watcher who has the regex pattern for that log message to match. Anytime the log messsage is changed the pattern has to be updated for the event to fire consistently across releases.
+
+Though this sounds like a redundant/roundabout way, this helps as below.
+- For III party code, not owned by SONiC, this is an acceptable solution
+- For code that SONiC own, a code update would take in the order of months to reach thousands of switches, but this approach can be applied instantly to every switch
+- A new event addition could be as simple as copying couple of small files to a switch and a rsyslog/container restart.
+
+##### Design at high level
+- Configure a rsyslog plugin as per process with rsyslog.d.
+- For logs raised by host processes, configure this plugin at host.
+- For logs raised by processes inside the container, configure for rsyslog.d running inside the container.
+- In this mode, there will be plugin processes running in host & in containers as one instance per process.
+- Provide the regex patterns to use for matching events as i/p to the plugin (*list of patterns for a process*).
+- Each plugin scans messasges **only** from a *single* process and matches only against patterns for that process.
+- For messages that match a pattern, retrieve parameters of interest and fire event using event reporter.
+- The rsyslog plugin binary, which does the parsing & reporting is a single binary in host, shared/used by all plugins.
+- The rsyslog plugin binary being under host control, ensures a single/unified behavior across all.
```

Comment:
>under host control [](http://example.com/codeflow?start=34&length=18)

Could you explain "under host control" ?

Reply:
The rsyslog-plugin binary is shared by all rsyslog based detectors. The detectors are running in host and multiple containers. If we do want a behavioral change, it is easy to accomplish.

Reply:
I'm a little confused about the plugin binary. How do the plugins communicate with this binary, is there anything that we would have to implement? If this binary does all the parsing & reporting does that mean this binary does the regex logic of creating events and sending them to the UDP listener? How would this binary talk to the UDP listener running in the telemetry container?